lib/testers/rule-tester.js

```diff
@@ -223,7 +251,21 @@ RuleTester.prototype = {
 
             validator.validate(config, "rule-tester");
 
-            return eslint.verify(code, config, filename);
+            // To cache AST.
+            eslint.reset();
+            eslint.on("Program", function(node) {
+                beforeAST = cloneDeeplyExcludesParent(node);
+
+                eslint.on("Program:exit", function(node) {
```

Comment:
Yes, almost the same.
But if it uses `Program` and `Program:exit`, RuleTester depends on an implementation that EventEmitter notifies events to listeners in the same order as added. And the tester must add the `Program:exit` listener after rules are set up. It's the reason for this line exists in the `Program` listener.

Ah... if rules use `onStart` / `onEnd`, these cannot resolve that...